jax/experimental/jax2iree/ir_builder.py

```diff
@@ -0,0 +1,102 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Optional, Sequence, Union
+
+import numpy as np
+
+from jax import core
+
+from .mlir_imports import *
+
+
+class Builder:
+  """An input module under construction."""
+
+  def __init__(self, context: Optional[ir.Context] = None):
+    self.context = context if context else ir.Context()
+    self.current_loc: Optional[ir.Location] = None
+    self.module = builtin.ModuleOp(loc=self.loc)
+    self.ip = ir.InsertionPoint(self.module.body)
+
+    self.context.allow_unregistered_dialects = True
+    mhlo.register_mhlo_dialect(self.context)
+    chlo.register_chlo_dialect(self.context)
+
+  @property
+  def loc(self) -> ir.Location:
+    if self.current_loc is None:
+      # TODO: Extract from traceback in some way.
+      return ir.Location.unknown(self.context)
+    return self.current_loc
+
+  def create_function(self, name: str, input_types: Sequence[ir.Type],
+                      return_types: Sequence[ir.Type]) -> "FunctionBuilder":
+    with self.context:
+      ftype = ir.FunctionType.get(input_types, return_types)
+      func_op = builtin.FuncOp(name, ftype, loc=self.loc, ip=self.ip)
+    return FunctionBuilder(self, func_op)
+
+  def convert_aval_to_ir_type(self, aval: core.AbstractValue) -> ir.Type:
+    with self.loc:
+      if isinstance(aval, core.ShapedArray):
+        element_type = self.convert_dtype_to_ir_type(aval.dtype)
+        shape = aval.shape  # TODO: Handle symbols, etc
+        return ir.RankedTensorType.get(shape, element_type)
+      elif isinstance(aval, core.UnshapedArray):
+        element_type = self.convert_dtype_to_ir_type(aval.dtype)
+        return ir.UnrankedTensorType.get(element_type)
+
+      raise NotImplementedError(f"Unsupported AbstractValue conversion: {aval}")
+
+  def convert_dtype_to_ir_type(self, dtype) -> ir.Type:
+    """Convert a dtype to an ir type."""
+    # TODO: Terrible.
```

Comment:
Haha, perhaps TODO: Switch on dtype and build appropriate type rather than just uniformly returning f32. f32 is purely for initial commit to create e2e thin slice.